src/ZkFold/Cardano/UPLC/Rollup.hs

```diff
@@ -1,56 +1,120 @@
+    -- ^ Update the rollup state using the proof.
+    | ForwardValidation
+    -- ^ Forwards validation of the rollup state update.
+    | CombineValue
+    -- ^ Combine the non-ada values locked in the rollup.
+    | AdjustStake
+    -- ^ Adjust the stake of the ada value locked in the rollup.
+    | UpgradeScript
+    -- ^ Update the script of the rollup to a new version.
+  deriving stock (Show, Generic)
+
+makeIsDataIndexed ''RollupRedeemer [('UpdateRollup,0),('ForwardValidation,1),('CombineValue,2),('AdjustStake,3),('UpgradeScript,4)]
+
+-- | Plutus script for verifying a rollup state transition.
 {-# INLINABLE rollup #-}
-rollup :: SetupBytes -> RollupRedeemer -> ScriptContext -> Bool
-rollup ledgerRules (RollupRedeemer proof addr val state update) ctx =
-        -- Verify the transition from the current state to the next state
-        verify @PlonkPlutus @HaskellCore ledgerRules state' proof
-        -- Check the current rollup output
-        && out  == TxOut addr val (OutputDatum $ Datum $ toBuiltinData state)  Nothing
-        -- Check the next rollup output
-        && out' == TxOut addr val (OutputDatum $ Datum $ toBuiltinData state') Nothing
-    where
-        -- Get the current rollup output
-        Just j = findOwnInput ctx
-        out    = txInInfoResolved j
-
-        -- Get the next rollup output
-        out'   = head $ txInfoOutputs $ scriptContextTxInfo ctx
-
-        -- Compute the next state
-        state' = toInput $ dataToBlake (state, update)
+rollup :: RollupSetup -> RollupRedeemer -> ScriptContext -> Bool
+rollup (RollupSetup ledgerRules dataCurrency threadValue feeAddress) (UpdateRollup proof update) ctx =
+  let
+    -- Get the current rollup output
+    out = txInInfoResolved $ case findOwnInput ctx of
+      Just j -> j
+      _      -> traceError "rollup: no input"
+
+    -- Get the address and state of the rollup
+    (addr, val, state) = case out of
+      TxOut a v (OutputDatum (Datum s)) _ -> (a, v, unsafeDataAsI s)
+      _                                   -> traceError "rollup: invalid redeemer"
+
+    -- Get state updates as token names of the data currency
+    update' =
+      map (unTokenName . fst) $
+      concatMap (toList . snd) $
+      filter (\(k, _) -> k == dataCurrency) $
+      -- Every referenced input must have the data currency as the second currency (the first one is ada).
+      map (head . tail . toList . getValue . txOutValue . txInInfoResolved)
+      (txInfoReferenceInputs $ scriptContextTxInfo ctx)
+
+    -- Get the next rollup output
+    out'   = head $ txInfoOutputs $ scriptContextTxInfo ctx
+
+    -- Get the fee output
+    outFee = head $ tail $ txInfoOutputs $ scriptContextTxInfo ctx
+    feeVal = getValue $ txOutValue outFee
+
+    -- Get bridge outputs
+    -- If the payment credential of the output coincides with the rollup payment credential, then this output transfers value to the rollup.
+    -- Otherwise, it transfers value from the rollup.
+    bridgeOutputs =
```

Comment:
I would have defined this filtering as
```haskell
bridgeOutputs =
  filter
    (\output -> case output of
        TxOut _ _ (OutputDatumHash _) Nothing -> True
        _                                     -> False)
    $ drop 3 $ txInfoOutputs $ scriptContextTxInfo ctx
```
because I am afraid that the filter lambda `\(TxOut _ _ (OutputDatumHash _) Nothing) -> True` is ambiguous when the pattern is not matched.  Do you think I am right?  (Also, I suggest `drop 3` instead of `tail . tail . tail`.)